File: Capítulo_7/Sets/Usando a classe Set/set.js
export class Set {
    #items;
    constructor() {
        this.#items = {};
    }

    // Metodo getter para #items
    _getItems() {
        return this.#items;
    }

    // Método setter para #items
    _setItems(item){
        this.#items[item] = item;
    }

    // Método has(element): devolve true se element estiver no conjunto, e false caso contrário.
    has(element) {
        return Object.prototype.hasOwnProperty.call(this._getItems(), element);
    }

    // add(element): adiciona um novo element ao conjunto.
    add(element) {
        if (!this.has(element)) {
            this._setItems(element);
            return true;
        }
        return false;
    }

    // Método delete(element): remove element do conjunto.
    delete(element) {
        if (this.has(element)) {
            delete this._getItems()[element];
            return true;
        }
        return false;
    }

    // Método clear(): remove todos os elementos do conjunto.
    clear() {
        this.#items = {};
    }

    // Método size(): devolve quantos elementos estão contidos no conjunto. É semelhante à propriedade length de um array.
    size() {
        let count = 0;
        for (let key in this._getItems()) {
            if (this._getItems().hasOwnProperty(key)) {
                count++;
            }
        }
        return count;
    }

    // Método values(): devolve um array com todos os valores (elementos) que estão no conjunto.
    values() {
        let values = [];
        for (let key in this._getItems()) {
            if (this._getItems().hasOwnProperty(key)) {
                values.push(this._getItems()[key]);
            }
        }
        return values;
    }
}

// const set = new Set();
// set.add(1);
// console.log(set.values()); // [1]
// console.log(set.has(1)); // true
// console.log(set.size()); // 1
// set.add(2);
// console.log(set.values()); // [1, 2]
// console.log(set.has(2)); // true
// console.log(set.size()); // 2
// set.delete(1);
// console.log(set.values()); // [2]
// set.delete(2);
// console.log(set.values()); // exibe []

File: Capítulo_7/Sets/Usando a classe Set/Código em Python/set.py
class Set:
    def __init__(self):
        self.__items = {}

    # Método getter para __items
    def get_items(self):
        return self.__items
    
    # Método setter para __items
    def set_items(self, item):
        self.__items[item] = item

    # Método has(element): devolve true se element estiver no conjunto, e false caso contrário.
    def has(self, element):
        return element in self.get_items()
    
    # add(element): adiciona um novo element ao conjunto.
    def add(self, element):
        if not self.has(element):
            self.set_items(element)
            return True
        return False
    
    # Método delete(element): remove element do conjunto.
    def delete(self, element):
        if self.has(element):
            del self.get_items()[element]
            return True
        return False
    
    # Método clear(): remove todos os elementos do conjunto.
    def clear(self):
        self.__items = {}

    # Método size(): devolve quantos elementos estão contidos no conjunto. É semelhante à propriedade length de um array.
    def size(self):
        count = 0
        for key in self.get_items():
            if self.has(key):
                count += 1
        return count
    
    # Método values(): devolve um array com todos os valores (elementos) que estão no conjunto.
    def values(self):
        values= []
        for key in self.get_items():
            if self.has(key):
                values.append(self.get_items()[key])
        return values

set = Set()
set.add(1)
print(set.values()) # [1]
print(set.has(1)) # True
print(set.size()) # 1
set.add(2)
print(set.values()) # [1, 2]
print(set.has(2)) # True
print(set.size()) # 2
set.delete(1)
print(set.values()) # [2]
set.delete(2)
print(set.size()) # 0
print(set.values()) # exibe []
set.add(1)
set.add(2)
set.add(3)
set.add(4)
print(set.values()) # [1, 2, 3, 4]
print(set.size()) # 4
set.clear()
print(set.size()) # 0